.gemini/styleguide.md

```diff
@@ -0,0 +1,397 @@
+# qsim style guide
+
+This style guide outlines the coding conventions for this project. There are
+separate subsections for Python, C++, and other file types below.
+
+## General guidance
+
+### Overall principles
+
+*   _Readability_: Code should be easy to understand for all team members.
+
+*   _Maintainability_: Code should be easy to modify and extend.
+
+*   _Consistency_: Adhering to a consistent style across all projects improves
+    collaboration and reduces errors.
+
+*   _Performance_: While readability is paramount, code should be efficient.
+
+### Overall development approach
+
+*   Use test-driven development (TDD) whenever possible. Write a failing test
+    _before_ writing new code. This ensures testability and clarifies
+    requirements.
+
+*   Isolate tests. tests must be independent and must not rely on the state of
+    other tests. Use setup and teardown functions to create a clean environment
+    for each test run.
+
+*   When new functions, classes, and files are introduced, they should also have
+    corresponding tests. Existing tests must continue to pass (or be updated)
+    when changes are introduced, and code should be covered by tests.
+
+*   Make sure to cover edge cases: Test for invalid inputs, null values, empty
+    arrays, zero values, and off-by-one errors.
+
+*   Use mocking for dependencies. In unit tests, external dependencies (e.g.,
+    databases, network services, file system) must be mocked to ensure the test
+    is isolated to the unit under test.
+
+*   Use asserts intelligently. Test assertions should be specific. Instead of
+    just asserting `true`, assert that a specific value equals an expected
+    value. Provide meaningful failure messages.
+
+*   Test coverage must be high. We don't require 100% coverage, but any
+    uncovered code must be annotated with `# pragma: no cover`. To ignore
+    coverage of a single line, place `# pragma: no cover` at the end of the
+    line. To ignore coverage for an entire block, start the block with a `#
+    pragma: no cover` comment on its own line.
+
+### Overall code format conventions
+
+This project generally follows Google coding conventions, with a few changes
+that are mostly defined in the various configuration files at the top level of
+the source tree. The following Google style guides are the starting points:
+
+*   [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html)
+
+*   [Google Python Style
+    Guide](https://google.github.io/styleguide/pyguide.html)
+
+*   [Google Markdown Style
+    Guide](https://google.github.io/styleguide/docguide/style.html)
+
+*   [Google Shell Style
+    Guide](https://google.github.io/styleguide/shellguide.html)
```

Comment:
Nit (disregard if original line breaks were from some markdown formatter) - for a plain text readability it is better to break the line before the URL value:
```suggestion
*   [Google Python Style Guide](
    https://google.github.io/styleguide/pyguide.html)

*   [Google Markdown Style Guide](
    https://google.github.io/styleguide/docguide/style.html)

*   [Google Shell Style Guide](
    https://google.github.io/styleguide/shellguide.html)
```

Please fix other similar instances below.